### PR TITLE
Resolving output shape ambiguity in transposed convolutions

### PR DESCRIPTION
**Pull request by davidtvs**
Adds an `output_shape` optional argument to `Conv2DTranspose` and `Conv3DTranspose`  to fix the issue described in #9883.

When `output_shape` is not specified the output shape is inferred (same as now); when specified, the output shape is passed directly to `keras.backend.conv2d_transpose` and `keras.backend.conv3d_transpose`.

### CODE REVIEW

**Comment by fchollet** [comment]
I think it would be a better API to specify an output padding in that case.

**Comment by davidtvs** [comment]
I personally prefer output shape as it's familiar and clearer. But looking at other frameworks, output padding is more common than output shape.

The best way to implement output padding would be to just pass it to [deconv_length](https://github.com/keras-team/keras/blob/master/keras/utils/conv_utils.py#L149) as a new argument and update the math, right?

**Comment by davidtvs** [comment]
Seems that the current implementation of `deconv_length` with `same` padding uses an equivalent output padding of 1 while `valid` and `full` use output padding 0:

Current implementation:
```py
>>> deconv_length(5, 2, 3, 'same')
10

>>> deconv_length(5, 2, 3, 'valid')
11
```

With output padding:
```py
>>> deconv_length(5, 2, 3, 'same', 0)
9
>>> deconv_length(5, 2, 3, 'same', 1)
10

>>> deconv_length(5, 2, 3, 'valid', 0)
11
```

How should we handle this so older code doesn't break?

**Comment by fchollet** [comment]
@davidtvs I would suggest defaulting `output_padding` to `None` (`output_padding=None` in signature) then documenting that `output_padding` will default to `1` with `same` and `0` otherwise (if unspecified, otherwise the specified value is used).

**Comment by fchollet** [comment]
Please also add more thorough unit testing to make sure that the statically inferred output shapes match the actual shapes being returned when running the computation graph on real data, across a wide range of parameters.

**Comment by davidtvs** [comment]
@fchollet upon further inspection, I found that the equivalent output padding I mentioned in my last comment can take other values other than 0 and 1 depending on the other parameters (stride, kernel size, etc). Therefore, I kept the current implementation which is used when `output_padding`is `None`, as you suggested; when it's not `None`, the following equation is used:

`out_dim = (in_dim - 1) * stride_size + kernel_size - 2 * pad + output_padding`

The tests all passed on my development branch, but here Travis is failing the build in `test_separable_conv_2d` with Python 3.6 and CNTK; I have no idea why, none of my code touches `test_separable_conv_2d`.

**Comment by fchollet** [comment]
Re-running tests.

**Comment by davidtvs** [comment]
I see it failed again, this time on `test_dtype` with both Python 2.7 and 3.6 and CNTK. @fchollet any clues as to what might be causing this?

**Comment by fchollet** [comment]
@souptc would you be able to advise with regard to this CNTK error? Thanks!

**Comment by souptc** [comment]
The failure is for the float16 type:

`def test_dtype(self):
        assert K.dtype(K.variable(1, dtype='float64')) == 'float64'
        assert K.dtype(K.variable(1, dtype='float32')) == 'float32'
        assert K.dtype(K.variable(1, dtype='float16')) == 'float16'`

As far as I know, the cntk 2.3 we  integrated in Keras does not have float16 support. I heard they add float16 support in recent version but not sure. Not sure how the float16 test passed on CNTK previously, could we skip float16 for cntk?

**Comment by souptc** [comment]
I saw someone skip this error in cntk_backend.py:
`def _convert_dtype_string(dtype):
    if dtype == np.float32:
        return 'float32'
    elif dtype == np.float64:
        return 'float64'
    elif dtype == np.float16:
        return 'float16'
`

@davidtvs , I think you need to merge with latest master?

**Comment by fchollet** [comment]
> I think you need to merge with latest master?

Re-running tests from master...

**Comment by fchollet** [comment]
Such passing, much green, wow.

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -649,6 +649,15 @@ class Conv2DTranspose(Conv2D):
             Specifying any stride value != 1 is incompatible with specifying
             any `dilation_rate` value != 1.
         padding: one of `"valid"` or `"same"` (case-insensitive).
+        output_padding: An integer or tuple/list of 2 integers,
+            specifying the amount of padding along the height and width
+            of the output tensor.
+            Can be a single integer to specify the same value for all
+            spatial dimensions.
+            The amount of output padding along a givern dimension must be
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Typo: "given"

File: keras/layers/convolutional.py [hunk 2]
```diff
@@ -743,6 +753,15 @@ def __init__(self, filters,
             **kwargs)
         self.input_spec = InputSpec(ndim=4)
 
+        self.output_padding = output_padding
+        if self.output_padding is not None:
+            self.output_padding = conv_utils.normalize_tuple(self.output_padding, 2, 'output_padding')
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Please shorten line (break after `(`)

File: keras/layers/convolutional.py [hunk 3]
```diff
@@ -743,6 +753,15 @@ def __init__(self, filters,
...
+            for stride, out_pad in zip(self.strides, self.output_padding):
+                if out_pad >= stride:
+                    raise ValueError('Stride ' + str(self.strides) + ' must be ' +
```
Comment (fchollet) [inline keras/layers/convolutional.py]: You don't need `+` to do string concatenation across two consecutive lines

File: keras/layers/convolutional.py [hunk 4]
```diff
@@ -871,6 +907,15 @@ class Conv3DTranspose(Conv3D):
             Specifying any stride value != 1 is incompatible with specifying
             any `dilation_rate` value != 1.
         padding: one of `"valid"` or `"same"` (case-insensitive).
+        output_padding: An integer or tuple/list of 3 integers,
+            specifying the amount of padding along the depth, height, and
+            width.
+            Can be a single integer to specify the same value for all
+            spatial dimensions.
+            The amount of output padding along a givern dimension must be
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Typo: "given"

File: keras/utils/conv_utils.py [hunk 5]
```diff
@@ -146,13 +146,41 @@ def conv_input_length(output_length, filter_size, padding, stride):
...
 
 
-def deconv_length(dim_size, stride_size, kernel_size, padding):
+def deconv_length(dim_size, stride_size, kernel_size, padding, output_padding):
+    """Determines output length of a transposed convolution given input length.
+
+    # Arguments
+        dim_size: integer.
+        stride_size: integer.
+        kernel_size: integer.
```
Comment (fchollet) [inline keras/utils/conv_utils.py]: Please add quick descriptions for the arguments, and capitalize the first character after `: `

File: keras/utils/conv_utils.py [hunk 6]
```diff
@@ -146,13 +146,41 @@ def conv_input_length(output_length, filter_size, padding, stride):
...
+            dim_size = dim_size * stride_size
+    else:
+        if padding == 'same':
+            pad = kernel_size // 2
+        elif padding == 'valid':
+            pad = 0
+        elif padding == 'full':
+            pad = kernel_size - 1
+
+        dim_size = (dim_size - 1) * stride_size + kernel_size - 2 * pad + output_padding
```
Comment (fchollet) [inline keras/utils/conv_utils.py]: Shorten line.

File: keras/layers/convolutional.py [hunk 7]
```diff
@@ -963,6 +1009,14 @@ def __init__(self, filters,
             bias_constraint=bias_constraint,
             **kwargs)
         self.input_spec = InputSpec(ndim=5)
+        self.output_padding = output_padding
+        if self.output_padding is not None:
+            self.output_padding = conv_utils.normalize_tuple(self.output_padding, 3, 'output_padding')
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Shorten line

File: keras/layers/convolutional.py [hunk 8]
```diff
@@ -963,6 +1009,14 @@ def __init__(self, filters,
...
+            for stride, out_pad in zip(self.strides, self.output_padding):
+                if out_pad >= stride:
+                    raise ValueError('Stride ' + str(self.strides) + ' must be ' +
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Same comment about `+`

File: tests/keras/layers/convolutional_test.py [hunk 9]
```diff
@@ -228,6 +232,24 @@ def test_conv2d_transpose():
                                                           padding=padding,
                                                           batch_input_shape=(None, None, 5, None))])
 
+    # Test invalid output padding for given stride. Output padding equal
+    # to stride
+    with pytest.raises(ValueError):
+        model = Sequential([convolutional.Conv2DTranspose(filters=filters,
```
Comment (fchollet) [inline tests/keras/layers/convolutional_test.py]: You can line break after `Conv2DTranspose(` to avoid overly long lines

**Comment by davidtvs** [comment]
Finally some green!

Thanks for your review @fchollet.

Do you think we should have the output shape equation in the `Conv2DTranspose` and `Conv3DTranspose` docstring? I think it would help understand what `output_padding` does exactly. Something like this:

```
# Output shape
        4D tensor with shape:
        `(batch, filters, new_rows, new_cols)`
        if `data_format` is `"channels_first"`
        or 4D tensor with shape:
        `(batch, new_rows, new_cols, filters)`
        if `data_format` is `"channels_last"`.
        `rows` and `cols` values might have changed due to padding.
        If `output_padding` is specified:
        `new_rows = (rows - 1) * strides[0] + kernel_size[0] - 2 * padding[0]
        + output_padding[0]`
        `new_cols = (cols - 1) * strides[1] + kernel_size[1] - 2 * padding[1]
        + output_padding[1]`
```

**Comment by fchollet** [comment]
Yes, that's valuable.

File: keras/layers/convolutional.py [hunk 10]
```diff
@@ -703,6 +711,9 @@ class Conv2DTranspose(Conv2D):
         `(batch, new_rows, new_cols, filters)`
         if `data_format` is `"channels_last"`.
         `rows` and `cols` values might have changed due to padding.
+        If `output_padding` is specified:<br/>
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Please do not include html markup in the docstrings. Two line breaks should be sufficient (or use markdown list formatting).

Comment (davidtvs) [reply keras/layers/convolutional.py]: I think two line breaks have too much empty space. A two line break at  "If `output_padding` is specified:" and the equations inside a code block looks better to me. What's your preference?

Comment (fchollet) [reply keras/layers/convolutional.py]: Docstrings should not contain HTML markup. Use plain markdown.
